### DISCUSSION TITLE
How to replace value on a prefilled form field in RHF?

### DISCUSSION THREAD

**Discussion by nonoumasy** [react-hook-form/react-hook-form#5094] [Q&A] [answered]
What I'm using?
React hook form + MUI
Firestore
use-places-autocomplete

I'm trying to update a form. First I fetch the data from firestore, setting that to state and then using that to prefill the form using RHF defaultValues. 
This works. However, when I try to add a new location, the form prefills it back with the defaultValues.
How to fix this? Any suggestions?

![ezgif com-gif-maker](https://user-images.githubusercontent.com/6312676/117562619-28e90280-b055-11eb-8044-6cbc8f693f23.gif)

You can also try it out on the app: https://hacktravels.com

PlacesAutocomplete component:
https://codesandbox.io/s/spring-night-u91zc?file=/src/PlacesAutocomplete.js

Edit component:
https://codesandbox.io/s/spring-night-u91zc?file=/src/Edit.js

Add component:
https://codesandbox.io/s/spring-night-u91zc?file=/src/Add.js

The code doesn't work. I just added it there for reference.
Adding something is not an issue. It's when I am editing the form.

**Comment by leapful** [comment]
@nonoumasy Please provide a working CSB to get more detail about the issue. Your current CSB doesn't show any example and also the steps to re-produce the issue.

**Reply by nonoumasy** [reply]
@leapful i'll try to edit the CSB with the minimal code. I started to recreate the issue but it got complicated. I'll try again hopefully soon.

**Comment by thanh-nguyen-95** [comment, marked as answer]
Since you did not provide a running example, here are my suggestions after looking at the "reference" code. I assume you're using the latest version of `react-hook-form`

- Use `<Controller />` or `useController` instead of using `register` directly. You can see the example here: https://react-hook-form.com/get-started#IntegratingwithUIlibraries
- You might need to call `setValue` (destructured from `useForm()`) manually when user select a location because your component might not emit `onChange` event when user do so
- Check the dependencies array of `useEffect`, `useCallback` or `useMemo`. They might memoize old value which causes the text not changing (very unlikely but better safe than sorry)

Hope this helps. Cheers!

**Reply by bluebill1049** [reply]
thank you very much @thanh-nguyen-95 for helping.

**Reply by nonoumasy** [reply]
thanks  @thanh-nguyen-95 . 
I'm using RHF "react-hook-form": "^6.11.0". I tried using the new version but had some issues with it and it was easier to just use pre-version 7.

The current form is not bad atleast when you add the first time, there are no issues. 
and when the form is edited, the only drawback is that the autocomplete isn't filling the form, however the User can still add a location to the input.

**Comment by nonoumasy** [comment]
@thanh-nguyen-95 
I was able to fix this issue up to a certain point.
RHF 6.11.1

I just have 2 issues left:
First issue: When I try to use the google autocomplete, I have to do it twice to fill it with the suggestion.
I think what's happening is that when i click on the suggestion the first time, it does another prefill cycle and then it doesnt do anymore and this allows the setValue change on the second attempt.
Second issue: After clicking submit, for a split second you see the original value on the input before existing the form. Not good UX.
Any thoughts?

```
import { useEffect } from 'react'

import usePlacesAutocomplete, {
    getGeocode,
    getLatLng,
} from "use-places-autocomplete";
import useOnclickOutside from "react-cool-onclickoutside";

import {
    TextField,
} from '@material-ui/core/';

const PlacesAutocomplete = ({ setLatitude, setLongitude, register, myLocation}) => {
    useEffect(() => {
        myLocation && setValue(myLocation)
    }, [])

    const {
        ready,
        value,
        suggestions: { status, data },
        setValue,
        clearSuggestions,
    } = usePlacesAutocomplete({
        requestOptions: {
            /* Define search scope here */
        },
        debounce: 300,
    });
    
    const ref = useOnclickOutside(() => {
        clearSuggestions();
    });

    const handleInput = (e) => {
        setValue(e.target.value);
    };

    const handleSelect = ({ description }) => () => {
        // When user selects a place, we can replace the keyword without request data from API
        // by setting the second parameter to "false"
        setValue(description, false);
        clearSuggestions();

        // Get latitude and longitude via utility functions
        getGeocode({ address: description })
            .then((results) => getLatLng(results[0]))
            .then(({ lat, lng }) => {
                console.log("📍 Coordinates: ", { lat, lng });
                setLatitude(lat)
                setLongitude(lng)
            })
            .catch((error) => {
                console.log("😱 Error: ", error);
            });
    };

    const renderSuggestions = () =>
        data.map((suggestion) => {
            const {
                place_id,
                structured_formatting: { main_text, secondary_text },
            } = suggestion;

            return (
                <li key={place_id} onClick={handleSelect(suggestion)}>
                    <strong>{main_text}</strong> <small>{secondary_text}</small>
                </li>
            );
        });

    return (
        <div ref={ref}>
            <TextField
                value={value}
                onChange={handleInput}
                disabled={!ready}
                variant="outlined"
                margin="normal"
                fullWidth
                id="location"
                label="location"
                name="location"
                type="text"
                inputRef={register}
                />
            {/* We can use the "status" to decide whether we should display the dropdown or not */}
            {status === "OK" && <ul>{renderSuggestions()}</ul>}
        </div>
    );
};

export default PlacesAutocomplete
```







https://user-images.githubusercontent.com/6312676/117895416-2cf26b80-b273-11eb-8468-674012dcb833.mov

**Reply by thanh-nguyen-95** [reply]
Based on your code snippet, my suggestions are:

- Use `<Controller>` or `useController` for your `TextField`
- Spend a litle bit of time to learn about _single source of truth_. Generally it means all your data should/must come from a single source (it's RHF data in your case). Your `TextField` is getting its `value` from `useAutoComplete` and not from RHF itself.

If I were you, this is how I would do it. Please keep in mind I'm not familiar with v6 so my syntax might not be correct

> Note: Every variables with `rhf` is from RHF
> Note 2: I only edit in places where I think matter the most, fill the rest however you need
> Note 3: Please use CodeSandBox or anything similar next time. It's hard (and takes long time) for us to help you without a running source code.

```jsx
function PlacesAutoComplete({ setLat, setLong, mylocation, rhfControl, setRhfValue }) {
  // setup auto complete hooks here

  useEffect(() => {
    if (mylocation) {
      setValue(mylocation)
      setRhfValue(mylocation)
    }
  }, [setRhfValue, setValue, mylocation])

  const handleSelect = ({ description }) => () => {
    // When user selects a place, we can replace the keyword without request data from API
    // by setting the second parameter to "false"
    setValue(description, false)
    clearSuggestions()

    // this might be the reason your form "flashes" before submit
    // add return here to prevent the API call below
    return

    // Get latitude and longitude via utility functions
    getGeocode({ address: description })
        .then((results) => getLatLng(results[0]))
        .then(({ lat, lng }) => {
            console.log("📍 Coordinates: ", { lat, lng })
            setLat(lat)
            setLong(lng)
        })
        .catch((error) => {
            console.log("😱 Error: ", error)
        });
  };

  return (
    <Controller 
      name="location"
      control={rhfControl}
      defaultValue={mylocation}
      render={({ name, ref, value: rhfValue, onChange }) => (
        <TextField
          name={name}
          inputRef={ref}
          value={rhfValue}
          onChange={(ev) => {
            const { target: { value } } = ev

            setValue(value)
            setRhfValue(value)
          }}
          /* other props here */
        />
      )}
    />
  )
}

export default PlacesAutoComplete
}
```

Edit: add more notes and update code snippet XD

**Reply by bluebill1049** [reply]
@thanh-nguyen-95 I really appreciated your support and help to everyone. any chance you can apply for the GitHub sponsor page? I would like to send some appreciation. you can talk to me here as well:  https://twitter.com/bluebill1049(if you use twitter)

**Reply by thanh-nguyen-95** [reply]
> @thanh-nguyen-95 I really appreciated your support and help to everyone. any chance you can apply for the GitHub sponsor page? I would like to send some appreciation. you can talk to me here as well: https://twitter.com/bluebill1049 (if you use twitter)

Thank you for the offer. Our company recently adopted `react-hook-form` and it's been a blast working with forms. Though there're ups and downs, it helps us alot. Thank you for creating such an amazing library ❤️

GitHub sponsor is not yet available in my country. Here's how I see it, the more time I spend on answering questions, the more time you have to improve the library. It's a win-win situation in my opinion.

**Reply by bluebill1049** [reply]
a huge thank you from me @thanh-nguyen-95 🙏 let me know once it's available. you deserve to be rewarded for your contribution to the community.
